petals-esb-deployer-impl/src/main/java/org/ow2/petals/deployer/runtimemodel/RuntimeContainer.java

```diff
@@ -109,17 +113,22 @@ public RuntimeServiceUnit getServiceUnit(final String id) {
      * @param serviceUnit
      *            must not be {code null}
      * @throws DuplicatedServiceUnitException
-     *             Service unit is already in the list
+     *             Service unit is already in the list and was not added
      */
     public void addServiceUnit(final RuntimeServiceUnit serviceUnit) throws DuplicatedServiceUnitException {
         assert serviceUnit != null;
-        if (serviceUnits.put(serviceUnit.getId(), serviceUnit) != null) {
+        String id = serviceUnit.getId();
+        if (serviceUnits.containsKey(id)) {
             throw new DuplicatedServiceUnitException("Service unit " + serviceUnit.getId() + " is already in the list");
         }
+        serviceUnits.put(id, serviceUnit);
     }
 
+    /**
+     * Adding or removing from the returned collection does not affect the container.
+     */
     public Collection<RuntimeServiceUnit> getServiceUnits() {
-        return serviceUnits.values();
+        return new HashSet<>(serviceUnits.values());
```

Comment:
I think you can find better using class `Collections.unmodifiableCollection()` to avoid adding and removing from the returned collection.